compiler/rustc_abi/src/layout.rs

```diff
@@ -901,36 +901,58 @@ pub trait LayoutCalculator {
-            assert!(field.is_sized());
-            align = align.max(field.align);
+            assert!(!field.is_unsized());
 
-            // If all non-ZST fields have the same ABI, forward this ABI
-            if optimize && !field.is_zst() {
-                // Discard valid range information and allow undef
-                let field_abi = match field.abi {
-                    Abi::Scalar(x) => Abi::Scalar(x.to_union()),
-                    Abi::ScalarPair(x, y) => Abi::ScalarPair(x.to_union(), y.to_union()),
-                    Abi::Vector { element: x, count } => {
-                        Abi::Vector { element: x.to_union(), count }
-                    }
-                    Abi::Uninhabited | Abi::Aggregate { .. } => Abi::Aggregate { sized: true },
-                };
+            if optimize {
+                // If all non-ZST fields have the same ABI, forward this ABI
+                if field.is_zst() {
+                    biggest_zst_align = biggest_zst_align.max(field.align);
+                } else {
+                    biggest_non_zst_align = biggest_non_zst_align.max(field.align);
+                    // Discard valid range information and allow undef
+                    let field_abi = match field.abi {
+                        Abi::Scalar(x) => Abi::Scalar(x.to_union()),
+                        Abi::ScalarPair(x, y) => Abi::ScalarPair(x.to_union(), y.to_union()),
+                        Abi::Vector { element: x, count } => {
+                            Abi::Vector { element: x.to_union(), count }
+                        }
+                        Abi::Uninhabited | Abi::Aggregate { .. } => Abi::Aggregate { sized: true },
+                    };
 
-                if size == Size::ZERO {
-                    // first non ZST: initialize 'abi'
-                    abi = field_abi;
-                } else if abi != field_abi {
-                    // different fields have different ABI: reset to Aggregate
-                    abi = Abi::Aggregate { sized: true };
+                    if let Some(abi) = &mut abi {
+                        if *abi != field_abi {
+                            // different fields have different ABI: reset to Aggregate
+                            *abi = Abi::Aggregate { sized: true };
+                        }
+                    } else {
+                        abi = Some(field_abi);
+                    }
                 }
             }
 
+            align = align.max(field.align);
             size = cmp::max(size, field.size);
         }
 
+        let abi = match abi {
+            None => Abi::Aggregate { sized: true },
+            Some(non_zst_abi) => {
+                if biggest_zst_align.abi > biggest_non_zst_align.abi {
+                    // If a zst has a bigger alignment than the non-zst fields,
+                    // we cannot use scalar layout, because scalar(pair)s can't be
+                    // more aligned than their primitive.
```

Comment:
This comment is misleading. `Abi::{Scalar,ScalarPair,Vector}` imply *fixed alignment* based on the their contents (and the `TargetDataLayout`). There's *zero* wiggle room (unless things changed again ofc).

(Also, you don't need `biggest_zst_align` at all either way - you can replace it with `align`, if `biggest_zst_align` is smaller than `align` would naturally be equal to `biggest_non_zst_align`)

Honestly there should be an e.g. `fn inherent_align(&self, cx: &C) -> Option<AbiAndPrefAlign>` method on `Abi` to query this properly for those 3 variants, it's getting a bit silly at this point...
(I believe @RalfJung's layout sanity checking code could also use that? I forget if it ended up duplicating some logic or not)

Reply:
For ScalarPair, my sanity check currently enforces `layout.layout.align().abi >= cmp::max(align1, align2)`, i.e. it leaves some wiggle room -- the type may have higher alignment than what the fields say. I am pretty sure I tried `==` and that did not pass the ui test suite.

Reply:
Maybe! I'm pretty sure we assume the exact alignment at least... somewhere?

Also, it's always an optimization, unlike the other two `Abi`s, so us using it less is not going to break anything *correctness-wise* (yes I have some PRs from this summer I need to get back to.... life caught up with me *sigh*).

Reply:
If https://github.com/rust-lang/rust/pull/105006 passes, the sanity check could also use such a `inherent_align` function, yeah.

Reply:
Oh wait actually, we currently also allow strengthening the alignment in vectors (i.e., the layout can have stricter alignment than the scalar type would require). And again probably I had a reason to allow that.

Reply:
https://github.com/rust-lang/rust/pull/105006 landed, so looks like you are right -- size and alignment of Scalar/ScalarPair/Vector types are entirely determined by their `Abi` (and target information).